README.md

```diff
@@ -28,21 +32,37 @@ Also check the [config.yaml documentation](https://github.com/haskell/ghcup-hs/b
 
 ## For contributors
 
-### Adding a new GHC version
+Most bindists are built downstream by the GHCup developers, e.g.:
+
+* https://github.com/stable-haskell/haskell-language-server
+* https://github.com/stable-haskell/cabal
+* https://github.com/stable-haskell/stack
+
+For GHC bindists there is no automation yet and it is done manually (e.g. for FreeBSD and Alpine i386).
+
+That makes contributions harder, because your pull requests to the metadata will be partial.
+
+## For upstream developers
+
+GHC, Cabal, HLS and stack developers distribute their own bindists via the `ghcup-vanilla-A.B.C.yaml` files. These
+files are primarily maintained by upstream developers at their own discretion. The GHCup project will perform no QA, fixes
+etc. on these bindists.
 
-1. open the latest `ghcup-<yaml-ver>.yaml`
-2. find the latest ghc version (in yaml tree e.g. `ghcupDownloads -> GHC -> 8.10.7`)
-3. copy-paste it
-4. adjust the version, tags, changelog, source url
-5. adjust the various bindist urls (make sure to also change the yaml anchors)
-6. run `cabal run ghcup-gen -- check             -f ghcup-<yaml-ver>.yaml`
-7. run `cabal run ghcup-gen -- check-tarballs    -f ghcup-<yaml-ver>.yaml -u 'ghc-8\.10\.8'`
-8. run `cabal run ghcup-gen -- generate-hls-ghcs -f ghcup-<yaml-ver>.yaml --format json -o hls-metadata-0.0.1.json`
-9. run `cabal run ghcup-gen -- generate-table    -f ghcup-<yaml-ver>.yaml --stdout` and adjust [docs/install](https://gitlab.haskell.org/haskell/ghcup-hs/-/blob/master/docs/install.md) tables
+This allows a clean separation between projects. Also see
+[GHCup is not an installer](https://hasufell.github.io/posts/2023-11-14-ghcup-is-not-an-installer.html).
+
+Updates to `ghcup-A.B.C.yaml` are carried out by the GHCup project.
+
+### Understanding tags
+
+Tags are documented [here](https://github.com/haskell/ghcup-hs/blob/master/lib/GHCup/Types.hs). Search for `data Tag`.
+Some tags are unique. Uniqueness is checked by `cabal run ghcup-gen -- check -f ghcup-<yaml-ver>.yaml`.
+
+If you want to check prereleases, do: `cabal run ghcup-gen -- check -f ghcup-prereleases-<yaml-ver>.yaml --channel=prerelease`
 
 ### During a pull request
 
-* make sure to always add new versions to both `ghcup-A.B.C.yaml` and `ghcup-vanilla-A.B.C.yaml`
+* For local testing see `cabal run ghcup-gen -- --help`
 * make sure to run the bindist action to check tool installation on all platforms: https://github.com/haskell/ghcup-metadata/actions/workflows/bindists.yaml
   - this is a manual pipeline
```

Comment:
I have the most trouble with this step: it seems to assume that I have permission to run workflows on this repository, doesn't it? It should be stated loud and clear, perhaps in a "Prerequisites" section: "In order to submit a PR for any channel, you need to have the permissions to run workflows on this repository". Maybe I'm misunderstanding and you expect me (someone who does not have the permissions) to create a fork, allow GtiHub Actions on the fork, and run the pipeline there? This is not trivial and should be spelled out.

Reply:
Yes, you need permissions. At least one developer of each upstream project has permissions.

From the cabal team that is @Kleidukos

I'm ok with adding another, but I'm reluctant to give out access like candy.

This is a sensitive repository with no safeguard between commits and the end user.

Reply:
> Why don't we come up with a less restrictive process? For example, a contributor may create a fork, allow actions on it, and run the pipeline.

Forks do not have access to private runners. It just doesn't work. And I'm not going to give access to every potential contributor. So I don't think there will be a less restrictive process.

A contributor can still provide a PR and I can assess it myself. This also has security reasons.

> Have you seen this: https://discourse.haskell.org/t/9124 ? The Cabal team is very liquid and is in a constant search of contributors, including release managers. If we want seamless and timely updates in the ecosystem, we must find more flexible ways to cooperate than giving access to one person once in a while through a process that is not documented anyhow.

I'm not sure what you are asking. I'm ok with giving 2 or maybe 3 cabal developers access and they can act as proxy for the release manager. I'm not going to give access to every possible release manager.

This is a trust based system. People have messed up before and I have revoked access before. People who have push access need a good understanding of how the metadata works. Yes, I can bombard you with documentation, but I question that will actually solve the issue. The issue is transferring of knowledge and responsive collaboration.

I am extremely responsive on this repo. A new contributor will have to create a couple of PRs before they will gain an understanding of the process. And I'm ok with that. You don't have to understand everything from the start.

Reply:
I appreciate your responsiveness! I was trying to do two things at once (maybe that's too much): (1) suggest ideas to improve the description in the README, so that people can understand enough to submit a PR, and (2) suggest improvements for the process. I see now that (2) is not feasible, that's fine. 

For (1), I can share my experience as the Cabal release manager for 3.12.1.0: I wasn't able to understand GHCup process for metadata updates enough to create a PR. That's after doing some research: reading readme and 	recent PRs. @geekosaur went a little further and opened a PR: https://github.com/haskell/ghcup-metadata/pull/225 But that PR was closed in a manner that I can only call abrupt. If that was me I'd be put off and wouldn't consider contributing to this repo anymore. I think it'd be good to improve the description of your process, so that if a PR has to be closed for some reason, you could spell out that reason with a reference to README.

Now, the reason I'm doing (1) is not getting permissions or anything like that. Instead, I want the next Cabal release manager (which may very well be a fresh person) not end up in the same place as @geekosaur and myself. That's why I'm trying to spell out assumptions that you seem to be making (as many as I can) and advocate for adding them to the README. More generally, I am trying to understand how the Cabal team may coordinate with ghcup more effectively. My thesis is that having a better description of the process will help tremendously.

On that note: if we could decide on a slight extension on the current process to **say explicitly in the README** that a PR may be submitted by anyone without the permissions to run the pipeline, but, since running the pipeline is an essential part of the process, the PR author should explicitly ask you to run the pipeline, that would be great.

What won't work, I think, is to pick a person X from every team, give them the permissions, and think that the whole team will magically understand the whole process. At least, after maintaining Cabal for a couple of years (by maintenance I mean updating various bits that need update to keep the lights on, like updating CI and stuff), I can reassure you: this is not a model that will work with the Cabal project as it stands. This is my personal opinion (as everything else I say), and other Cabal team members can disagree.

Reply:
> That's after doing some research: reading readme and recent PRs. @geekosaur went a little further and opened a PR: https://github.com/haskell/ghcup-metadata/pull/225 But that PR was closed in a manner that I can only call abrupt. If that was me I'd be put off and wouldn't consider contributing to this repo anymore.

I'm confused.

I used @geekosaur's PR and created a follow-up that ran all the required pipelines: https://github.com/haskell/ghcup-metadata/pull/226

It was merged. The contribution was successful. I did all the outstanding work for you.

What exactly do you have a problem with?

> What won't work, I think, is to pick a person X from every team, give them the permissions, and think that the whole team will magically understand the whole process. At least, after maintaining Cabal for a couple of years (by maintenance I mean updating various bits that need update to keep the lights on, like updating CI and stuff), I can reassure you: this is not a model that will work with the Cabal project as it stands. This is my personal opinion (as everything else I say), and other Cabal team members can disagree.

GHCup is a *distribution*.

The process to become a core member of a distribution is usually extremely tedious. If you want to submit a package to debian as a casual contributor, you have to jump through a lot of hoops. That is on purpose.

And yet, I have given several members of the Haskell core tooling direct access to this repo and allowed them to push to the vanilla channel.

I will **not** give every cabal contributor (or release manager) access to this repository.

----

Distributing the knowledge of how the metadata works is a perfectly reasonable goal. That is why I created this PR. But at the moment I'm not sure whether documentation is the right way to distribute that knowledge.

I also gave a talk at ZuriHac about GHCup. The slides are available here: https://www.haskell.org/ghcup/talks/Zurihac2024/GHCup.html

Reply:
> Oh, I didn't notice that one! It may help to leave a note about a superceeding PR on the thread for the initial one because GitHub doesn't say email notifications about cross-thread references (you referenced the initial PR in yours but no one got notified about that).

He did say he was going to do that, because workflows don't work on PRs originating in other repos for whatever reason.

BTW, I was unable to find a way to validate my metadata file (`ghcup-gen` apparently wasn't it, since it didn't notice that I'd inadvertently duplicated a bunch of tags). Is there a way to do this before submitting a PR?

Reply:
> BTW, I was unable to find a way to validate my metadata file (ghcup-gen apparently wasn't it, since it didn't notice that I'd inadvertently duplicated a bunch of tags). Is there a way to do this before submitting a PR?

`ghcup-gen check` does check for duplicated tags:

https://github.com/haskell/ghcup-metadata/blob/develop/ghcup-gen/Validate.hs#L135-L165

Reply:
> Noted. This wasn't something I asked for though. All that I ask for is a description of the process with enough details for some Cabal dev to follow. 

Tbh, I feel like I'm losing energy on this topic.

I suggest you simply try to contribute. Go through:

- the documentation: https://www.haskell.org/ghcup/dev/
- the zurihac slides
- the readme in this repo
- previous PRs
- the code of ghcup-gen (it is really small)

If you still have specific questions afterwards, feel free to ask. Also feel free to fix the documentation/take up this PR.

Reply:
Sounds good! Thanks for the insightful discussion: it will help future generations of Cabal release managers! Please, make sure to merge this PR. And keeping the README up to date would be going above and beyond! Thanks again!